src/node.c

```diff
@@ -150,6 +155,15 @@ cmark_node_type cmark_node_get_type(cmark_node *node) {
   }
 }
 
+int cmark_node_set_type(cmark_node * node, cmark_node_type type) {
+  if (type == node->type)
+    return true;
+
+  free_node_as(node);
+  node->type = type;
+  return true;
+}
```

Comment:
> check whether the new node type is valid for the node's parent.

Indeed, that would be a valid case for returning false in there, will do

> initialize the node to its default state (unless the content can be reused).

Note that this function does (re)initialize the node to its initial type, that's what "free_node_as" is doing, in case you're not aware node types are implemented as a union.

> Maybe the parser could first assign an internal node type like UNKNOWN_BLOCK which can later be changed to a known block type?

I'm not sure why we'd want to do that? Note that the intended usage for this function is also to modify a previously defined node type, as explained in the message for the commit introducing this API (https://github.com/MathieuDuponchelle/cmark/commits/extensions)